src/backend/distributed/connection/placement_connection.c

```diff
@@ -229,272 +222,373 @@ GetPlacementConnection(uint32 flags, ShardPlacement *placement, const char *user
-		returnConnectionReference->userName =
-			MemoryContextStrdup(TopTransactionContext, userName);
-		dlist_push_tail(&placementEntry->connectionReferences,
-						&returnConnectionReference->placementNode);
-
-		/* record association with shard, for invalidation */
-		AssociatePlacementWithShard(placementEntry, placement);
-
-		/* record association with connection, to handle connection closure */
-		dlist_push_tail(&connection->referencedPlacements,
-						&returnConnectionReference->connectionNode);
-	}
+	placementAccess->placement = placement;
 
 	if (flags & FOR_DDL)
 	{
-		placementEntry->modifyingConnection = returnConnectionReference;
-		returnConnectionReference->hadDDL = true;
+		placementAccess->accessType = PLACEMENT_ACCESS_DDL;
 	}
-	if (flags & FOR_DML)
+	else if (flags & FOR_DML)
 	{
-		placementEntry->modifyingConnection = returnConnectionReference;
-		returnConnectionReference->hadDML = true;
+		placementAccess->accessType = PLACEMENT_ACCESS_DML;
 	}
-
-	if (freeUserName)
+	else
 	{
-		pfree(freeUserName);
+		placementAccess->accessType = PLACEMENT_ACCESS_SELECT;
 	}
 
-	return returnConnectionReference->connection;
+	return StartPlacementListConnection(flags, list_make1(placementAccess), userName);
 }
 
 
 /*
- * CheckExistingPlacementConnections check whether any of the existing
- * connections is usable. If so, return it, otherwise return NULL.
+ * GetPlacementListConnection establishes a connection for a set of placement
+ * accesses.
  *
- * A connection is usable if it is not in use, the user matches, DDL/DML usage
- * match and cached connection are allowed.  If no existing connection
- * matches, but a new connection would conflict (e.g. a connection already
- * exists but isn't usable, and the desired connection needs to execute
- * DML/DML) an error is thrown.
+ * See StartPlacementListConnection for details.
+ */
+MultiConnection *
+GetPlacementListConnection(uint32 flags, List *placementAccessList, const char *userName)
+{
+	MultiConnection *connection = StartPlacementListConnection(flags, placementAccessList,
+															   userName);
+
+	FinishConnectionEstablishment(connection);
+	return connection;
+}
+
+
+/*
+ * StartPlacementListConnection returns a connection to a remote node suitable for
+ * a placement accesses (SELECT, DML, DDL) or throws an error if no suitable
+ * connection can be established if would cause a self-deadlock or consistency
+ * violation.
  */
-static ConnectionReference *
-CheckExistingPlacementConnections(uint32 flags,
-								  ConnectionPlacementHashEntry *placementEntry,
-								  const char *userName)
+MultiConnection *
+StartPlacementListConnection(uint32 flags, List *placementAccessList,
+							 const char *userName)
 {
-	dlist_iter it;
+	char *freeUserName = NULL;
+	bool foundModifyingConnection = false;
+	ListCell *placementAccessCell = NULL;
+	List *placementEntryList = NIL;
+	ListCell *placementEntryCell = NULL;
+	MultiConnection *connection = NULL;
```

Comment:
`connection` is not so much specific to 1 particular placement, but is to be used for all placements in the list, unlike `placementConnection`. Maybe something like `chosenConnection`? Could also do `placementListConnection` maybe.